### PR TITLE
Change the way how start and end timestamps are converted to frames

### PR DESCRIPTION
**Pull request by JanuszL**
- by default VideoReader file_list timestamps are converted to
 `ceil(start_time_stamp * FPS)` and `floor(end_time_stamp * FPS)`.
  This PR adds an option `file_list_include_preceding_frame` to change
  the behavior to `floor(start_time_stamp * FPS)` and
  `ceil(end_time_stamp * FPS)` respectively. As the sequence includes
  the first frame and excludes the last it is more aligned with the
  the way how visible display time is connected with displayed frames

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [x] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
- by default VideoReader file_list timestamps are converted to
 `ceil(start_time_stamp * FPS)` and `floor(end_time_stamp * FPS)`.
  This PR adds an option `file_list_include_preceding_frame` to change
  the behavior to `floor(start_time_stamp * FPS)` and
  `ceil(end_time_stamp * FPS)` respectively. As the sequence includes
  the first frame and excludes the last it is more aligned with the
  the way how visible display time is connected with displayed frames

#### Additional information
- Affected modules and functionalities:
  - video reader

- Key points relevant for the review:
  - logic and test

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**:  RVID.22, RVID.23, RVID.24

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/loader/video_loader.h [hunk 1]
```diff
@@ -163,13 +163,20 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
...
       stats_({0, 0, 0, 0, 0}),
       current_frame_idx_(-1),
       stop_(false) {
     DALI_ENFORCE(stride_ > 0, "Stride should be > 0");
     if (step_ < 0)
       step_ = count_ * stride_;
+    if (!file_list_include_preceding_frame_) {
+      DALI_WARN("`file_list_include_preceding_frame_` is set to false (or not set at all). In "
+                "future releases, the default behavior would be changed to True and the argument "
+                "will be removed.");
```
Comment (klecki) [inline dali/operators/reader/loader/video_loader.h:178]: Adding mandatory argument that will be removed in future releases is a pathway to the code compatibility issues between releases.

Comment (JanuszL) [reply dali/operators/reader/loader/video_loader.h:178]: Maybe we can change the default first and then remove?

File: dali/operators/reader/video_reader_op.cc [hunk 2]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
 the end frame number will be rounded down.
 
 Frame numbers start from 0.)code", false)
+  .AddOptionalArg("file_list_include_preceding_frame",
+      R"code(Changes the behavior how `file_list` start and end frame timestamps are translated
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:165]:
```suggestion
      R"code(Changes the behavior about how ``file_list`` start and end frame timestamps are translated
```

File: dali/operators/reader/video_reader_op.cc [hunk 3]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+to the frame number.
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:166]:
```suggestion
to a frame number.
```

File: dali/operators/reader/video_reader_op.cc [hunk 4]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+
+If the start/end timestamps are provided in file_list as timestamps, the start frame is
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:168]:
```suggestion
If the start/end timestamps are provided in ``file_list`` as timestamps, the start frame is
```

File: dali/operators/reader/video_reader_op.cc [hunk 5]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+calculated as `ceil(start_time_stamp * FPS)` and the end as `floor(end_time_stamp * FPS)`.
+If this option is turned on the equation changes to `floor(start_time_stamp * FPS)` and
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:170]:
```suggestion
If this argument is set to True, the equation changes to ``floor(start_time_stamp * FPS)`` and
```

File: dali/operators/reader/video_reader_op.cc [hunk 6]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
 the end frame number will be rounded down.
 
 Frame numbers start from 0.)code", false)
+  .AddOptionalArg("file_list_include_preceding_frame",
+      R"code(Changes the behavior how `file_list` start and end frame timestamps are translated
+to the frame number.
+
+If the start/end timestamps are provided in file_list as timestamps, the start frame is
+calculated as `ceil(start_time_stamp * FPS)` and the end as `floor(end_time_stamp * FPS)`.
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:169]:
```suggestion
calculated as ``ceil(start_time_stamp * FPS)`` and the end as ``floor(end_time_stamp * FPS)``.
```

File: dali/operators/reader/video_reader_op.cc [hunk 7]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+If this option is turned on the equation changes to `floor(start_time_stamp * FPS)` and
+`ceil(end_time_stamp * FPS)` respectively. In effect, the first returned frame is no later, and
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:171]:
```suggestion
``ceil(end_time_stamp * FPS)`` respectively. In effect, the first returned frame is not later, and
```

File: dali/operators/reader/video_reader_op.cc [hunk 8]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+the end frame no earlier than provided timestamps what is more aligned with how the visible
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:172]:
```suggestion
the end frame not earlier, than the provided timestamps. This behavior is more aligned with how the visible
```

File: dali/operators/reader/video_reader_op.cc [hunk 9]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+timestamps are correlated with displayed video frames.
+
+.. note::
+
+  When `file_list_frame_num` is set this option doesn't take any effect.
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:177]:
```suggestion
  When ``file_list_frame_num`` is set to True, this option doesn't take any effect.
```

File: dali/operators/reader/video_reader_op.cc [hunk 10]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+
+.. warning::
+
+  This option is meant to preserve the legacy behavior. It will be removed in the following
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:181]:
```suggestion
  This option is available for legacy behavior compatibility, and will be removed in the following
```

File: dali/operators/reader/video_reader_op.cc [hunk 11]
```diff
@@ -161,6 +161,26 @@ If floating point values have been provided, the start frame number will be roun
...
+  releases and the default behavior would be changed to ``file_list_include_preceding_frame=True``.
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.cc:182]:
```suggestion
  releases, being the default behavior equivalent to ``file_list_include_preceding_frame=True``.
```

File: dali/operators/reader/loader/video_loader.h [hunk 12]
```diff
@@ -163,13 +163,20 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
...
+                                                          "file_list_include_preceding_frame")),
       pad_sequences_(spec.GetArgument<bool>("pad_sequences")),
       stats_({0, 0, 0, 0, 0}),
       current_frame_idx_(-1),
       stop_(false) {
     DALI_ENFORCE(stride_ > 0, "Stride should be > 0");
     if (step_ < 0)
       step_ = count_ * stride_;
+    if (!file_list_include_preceding_frame_) {
+      DALI_WARN("`file_list_include_preceding_frame_` is set to false (or not set at all). In "
```
Comment (jantonguirao) [inline dali/operators/reader/loader/video_loader.h:176]:
```suggestion
      DALI_WARN("``file_list_include_preceding_frame`` is set to False (or not set at all). In "
```

File: dali/operators/reader/loader/video_loader.h (hunk 1 above)
Comment (klecki) [reply dali/operators/reader/loader/video_loader.h:178]: The main point is, that this creates an opportunity to write code, that will be at that point the way advertised by us (you need to set `file_list_include_preceding_frame_` to True), but removing that argument will lead to it not working in the future (error on setting argument that doesn't exist). I'm not convinced this is the best way of doing stuff.

File: dali/operators/reader/loader/video_loader.h (hunk 12 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:176]: Done

File: dali/operators/reader/loader/video_loader.h (hunk 1 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:178]: Done as discussed

File: dali/operators/reader/video_reader_op.cc (hunk 2 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:165]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 3 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:166]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 6 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:169]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 5 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:170]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 7 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:171]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 9 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:177]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 10 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc:181]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 11 above)
Comment (awolant) [reply dali/operators/reader/video_reader_op.cc]: Done, whole sentence was removed.

**Comment by awolant** [comment]
!build

File: dali/operators/reader/loader/video_loader.h [hunk 13]
```diff
@@ -163,13 +163,19 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
...
       pad_sequences_(spec.GetArgument<bool>("pad_sequences")),
       stats_({0, 0, 0, 0, 0}),
       current_frame_idx_(-1),
       stop_(false) {
     DALI_ENFORCE(stride_ > 0, "Stride should be > 0");
     if (step_ < 0)
       step_ = count_ * stride_;
+    if (!file_list_include_preceding_frame_) {
+      DALI_WARN("``file_list_include_preceding_frame_`` is set to False (or not set at all). In "
+                "future releases, the default behavior would be changed to True..");
```
Comment (klecki) [inline dali/operators/reader/loader/video_loader.h:177]:
```suggestion
                "future releases, the default behavior would be changed to True.");
```

File: dali/operators/reader/loader/video_loader.h [hunk 14]
```diff
@@ -163,13 +163,19 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
...
+                                                          "file_list_include_preceding_frame")),
       pad_sequences_(spec.GetArgument<bool>("pad_sequences")),
       stats_({0, 0, 0, 0, 0}),
       current_frame_idx_(-1),
       stop_(false) {
     DALI_ENFORCE(stride_ > 0, "Stride should be > 0");
     if (step_ < 0)
       step_ = count_ * stride_;
+    if (!file_list_include_preceding_frame_) {
+      DALI_WARN("``file_list_include_preceding_frame_`` is set to False (or not set at all). In "
```
Comment (klecki) [inline dali/operators/reader/loader/video_loader.h:176]:
```suggestion
      DALI_WARN("`file_list_include_preceding_frame` is set to False (or not set at all). In "
```

File: dali/operators/reader/loader/video_loader.h (hunk 13 above)
Comment (klecki) [reply dali/operators/reader/loader/video_loader.h:177]: Can we guestimate when we will do it, and put it here?

File: dali/operators/reader/loader/video_loader.h [hunk 15]
```diff
@@ -245,15 +251,33 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
         } else {
           auto frame_rate = av_inv_q(file.frame_base_);
           if (start >= 0) {
-            start_frame = static_cast<int>(std::ceil(start * av_q2d(frame_rate)));
+            if (file_list_include_preceding_frame_) {
```
Comment (klecki) [inline dali/operators/reader/loader/video_loader.h:254]: Can you make those if consistent?
Once it's `if (file_list_include_preceding_frame_)` on top level, the other time it's the nested one.

File: dali/operators/reader/loader/video_loader.h [hunk 16]
```diff
@@ -163,13 +163,19 @@ class VideoLoader : public Loader<GPUBackend, SequenceWrapper> {
       codec_id_(0),
       skip_vfr_check_(spec.GetArgument<bool>("skip_vfr_check")),
       file_list_frame_num_(spec.GetArgument<bool>("file_list_frame_num")),
+      file_list_include_preceding_frame_(spec.GetArgument<bool>(
```
Comment (jantonguirao) [inline dali/operators/reader/loader/video_loader.h:166]: nitpick: indentation seems off here

**Review by jantonguirao — APPROVED** [review]

**Review by klecki — COMMENTED** [review]
Minor issues.

File: dali/operators/reader/loader/video_loader.h (hunk 16 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:166]: Done

File: dali/operators/reader/loader/video_loader.h (hunk 14 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:176]: Done

File: dali/operators/reader/loader/video_loader.h (hunk 13 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:177]: I don know tbh. And I prefer not to make promises, if we are not sure.

File: dali/operators/reader/loader/video_loader.h (hunk 15 above)
Comment (awolant) [reply dali/operators/reader/loader/video_loader.h:254]: Done

**Review by klecki — APPROVED** [review]

**Comment by awolant** [comment]
!build

**Pull request merged**